server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java

```diff
@@ -427,6 +428,10 @@ public void seqID(SeqNoFieldMapper.SequenceIDFields seqID) {
 
         @Override
         public void addDynamicMapper(Mapper mapper) {
+            // eagerly check field name limit here to avoid OOM errors
+            if (mappingLookup.getMapper(mapper.name()) == null && newFieldsSeen.add(mapper.name())) {
```

Comment:
tricky bits: what happens when we dynamically add an object field? and what is the scenario where the mapper is not found (meaning it's a new one?) but is already in the set?

Reply:
- object fields are not treated specially in this logic (it's counted as a proper field, similar as it's done in the field limit check)
- the scenario where the mapper is not found but is already in the set covers scenarios like the one found in the `testLimitNestedDocsDefaultSettings` test:
```
"foo" : [ { "bar" : 1 }, { "bar" : 2 }, { "bar" : 3} ]
```
If the "bar" field has not been mapped prior to receiving this document, parsing this doc results in 3 separate calls to `addDynamicMapper`. Ultimately these are merged together in `DocumentParser.createDynamicUpdate`, so should only count as 1 (and will only count as 1 for the fields limit check that is run on the merged mapping). If we didn't have this logic, we would now start failing the fields limit check at parse time because we would count these as separate fields even though it's just one.

Reply:
makes sense, I had forgotten that we have the same mapper multiple times, and we merge those to prevent conflicts between them. 

My question around object fields revolved around the fact that mappingLookup#getMapper does not seem expose object mappers, so I was wondering about the behaviour when an object with the provided name is defined. I guess that the check will always return null hence we may count objects that are already mapped as new? I may be missing something but maybe it is worth adding a test for this scenario to verify what happens?

Reply:
oh well maybe that never happens because we only call addDynamicMapper for fields that are not known... still may make sense to verify that we are doing the right thing.

Reply:
I guess that my next question is why we need the getMapper call. When are we calling addDynamicMapper providing a mapper that mapping lookup is already aware of?

Reply:
> When are we calling addDynamicMapper providing a mapper that mapping lookup is already aware of?

I wasn't sure how much we can rely on that. For ObjectMapper I felt that this would not be an issue, but for other arbitrary FieldMapper (which can be added by plugins etc.) I was wondering if the code should have an additional guard against that (e.g. ConstantKeywordFieldMapper directly calls this, and even though the guard is not necessary for this class, it's unclear to me everyone would stick to this contract). If we have for example a new enum field type that would allow values to be populated dynamically, a dynamic mapper would possibly want to add a new enum value to existing enum field (i.e. dynamic mapping update to the existing field).

Reply:
oh boy! you are right that we can't rely on it. I find it pretty scary and surprising that field mappers can add dynamic mappers themselves, I was only looking at the ordinary flow where this method is called when parsing a document because the field is not already mapped. Could you add a small comment that explains why we have that if, and possibly a test that fails if we remove it? I think that you could use constant keyword in your test to show how an already mapped field can be among the dynamic mappers yet it should not be counted. Thanks for going through this with me :)

Reply:
I've pushed a commit (e596aea) that adds comments to the code, as well as a test for constant_keyword (that fails when you remove the extra check whether the field already exists). I've also expanded the check to cover object mappers. Even though I can't construct such a test case, I would like to be extra careful in this place.

Reply:
Thanks! if you really wanted to test also the object mapper scenario I think you could define a custom field that as part of its parseCreateField adds a dynamic mapper for an object field. It is not required, it sounds like a crazy usecase and stuff we may look at in the future and wonder why we tested it.

Reply:
yes, I find it more confusing to add such a synthetic test.